File: app/src/main/java/com/mercandalli/android/apps/files/file_list/ScaleAnimationAdapter.java
package com.mercandalli.android.apps.files.file_list;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.annotation.TargetApi;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.Interpolator;

/**
 * {@link RecyclerView} adapter with animation.
 */
@TargetApi(Build.VERSION_CODES.HONEYCOMB)
public class ScaleAnimationAdapter extends RecyclerView.Adapter<ViewHolder> implements
        Animator.AnimatorListener {

    /**
     * The {@link RecyclerView.Adapter} with animated {@link ViewHolder}.
     */
    @NonNull
    private final RecyclerView.Adapter mAdapter;

    /**
     * The {@link RecyclerView} with animated {@link ViewHolder}.
     */
    @NonNull
    private final RecyclerView mRecyclerView;

    /**
     * The animation duration for each {@link RecyclerView}.
     */
    private int mDuration = 220;

    /**
     * The animation offset duration between each {@link RecyclerView}.
     */
    private int mOffsetDuration = 50;

    /**
     * The animation {@link Interpolator}.
     */
    @NonNull
    private Interpolator mInterpolator = new AccelerateDecelerateInterpolator();

    /**
     * The number of {@link ViewHolder} items per line.
     */
    private int mViewHolderPerLine;

    private int mLastPosition = -1;
    private boolean mIsFirstOnly = false;
    private int mCounter;
    private boolean mAnimsInitialized;
    private final float mFrom;

    @Nullable
    private NoAnimatedPosition mNoAnimatedPosition;

    public ScaleAnimationAdapter(RecyclerView recyclerView, RecyclerView.Adapter adapter) {
        this(recyclerView, adapter, 1, 0.0F);
    }

    public ScaleAnimationAdapter(RecyclerView recyclerView, RecyclerView.Adapter adapter, int cardPerLine) {
        this(recyclerView, adapter, cardPerLine, 0.0F);
    }

    public ScaleAnimationAdapter(
            @NonNull final RecyclerView recyclerView,
            @NonNull final RecyclerView.Adapter adapter,
            int cardPerLine,
            float from) {
        setHasStableIds(true);
        mRecyclerView = recyclerView;
        mAdapter = adapter;
        mViewHolderPerLine = cardPerLine;
        mFrom = from;
        adapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onChanged() {
                mLastPosition = -1;
                notifyDataSetChanged();
            }

            @Override
            public void onItemRangeChanged(int positionStart, int itemCount) {
                notifyItemRangeChanged(positionStart, itemCount);
            }

            @Override
            public void onItemRangeChanged(int positionStart, int itemCount, Object payload) {
                notifyItemRangeChanged(positionStart, itemCount, payload);
            }

            @Override
            public void onItemRangeInserted(int positionStart, int itemCount) {
                notifyItemRangeInserted(positionStart, itemCount);
            }

            @Override
            public void onItemRangeRemoved(int positionStart, int itemCount) {
                notifyItemRangeRemoved(positionStart, itemCount);
            }

            @Override
            public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
                notifyDataSetChanged();
            }
        });
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return this.mAdapter.onCreateViewHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        this.mAdapter.onBindViewHolder(holder, position);
        if (this.mIsFirstOnly && position <= this.mLastPosition) {
            clear(holder.itemView);
        } else if (mNoAnimatedPosition == null || mNoAnimatedPosition.isAnimatedItem(position)) {
            final Animator[] animators = this.getAnimators(holder.itemView);
            if (!mAnimsInitialized) {
                for (Animator anim : animators) {
                    anim.setDuration((long) this.mDuration);
                    anim.setInterpolator(this.mInterpolator);
                    increaseCounter();

                    if (mViewHolderPerLine > 1 && position >= mViewHolderPerLine) {
                        anim.setStartDelay(mOffsetDuration * (position - mViewHolderPerLine / 2));
                    } else {
                        anim.setStartDelay(mOffsetDuration * position);
                    }
                    anim.addListener(this);
                    anim.start();
                }
            } else {
                holder.itemView.setScaleX(1);
                holder.itemView.setScaleY(1);
            }

            this.mLastPosition = position;
        } else {
            holder.itemView.setScaleX(1);
            holder.itemView.setScaleY(1);
            this.mLastPosition = position;
        }
    }

    @Override
    public long getItemId(int position) {
        return mAdapter.getItemId(position);
    }

    @Override
    public void onAnimationStart(Animator animation) {

    }

    @Override
    public void onAnimationEnd(Animator animation) {
        if (decreaseCounter()) {
            mAnimsInitialized = true;
        }
    }

    @Override
    public void onAnimationCancel(Animator animation) {

    }

    @Override
    public void onAnimationRepeat(Animator animation) {

    }

    public synchronized int getCounter() {
        return mCounter;
    }

    public synchronized void increaseCounter() {
        mCounter++;
    }

    public int getItemCount() {
        return this.mAdapter.getItemCount();
    }

    public void setDuration(int duration) {
        this.mDuration = duration;
    }

    public void setOffsetDuration(int mOffsetDuration) {
        this.mOffsetDuration = mOffsetDuration;
    }

    public void setInterpolator(@NonNull final Interpolator interpolator) {
        this.mInterpolator = interpolator;
    }

    public void setStartPosition(int start) {
        this.mLastPosition = start;
    }

    private RecyclerView getRecyclerView() {
        return mRecyclerView;
    }

    public int getItemViewType(int position) {
        return this.mAdapter.getItemViewType(position);
    }

    private synchronized boolean decreaseCounter() {
        mCounter--;
        return mCounter == 0;
    }

    private Animator[] getAnimators(View view) {
        view.setScaleX(0);
        view.setScaleY(0);
        return new Animator[]{ObjectAnimator.ofFloat(view, "scaleX", this.mFrom, 1.0F), ObjectAnimator.ofFloat(view, "scaleY", this.mFrom, 1.0F)};
    }

    private static void clear(@NonNull final View v) {
        ViewCompat.setAlpha(v, 1.0F);
        ViewCompat.setScaleY(v, 1.0F);
        ViewCompat.setScaleX(v, 1.0F);
        v.setPivotY((float) (v.getMeasuredHeight() / 2));
        ViewCompat.setPivotX(v, (float) (v.getMeasuredWidth() / 2));
        ViewCompat.animate(v).setInterpolator(null);
    }

    public void setNoAnimatedPosition(@Nullable final NoAnimatedPosition noAnimatedPosition) {
        mNoAnimatedPosition = noAnimatedPosition;
    }

    public void reset() {
        mLastPosition = -1;
        mAnimsInitialized = false;
    }

    public interface NoAnimatedPosition {

        /**
         * The the item not animated.
         *
         * @param position The position to check.
         * @return True if the item is animated.
         */
        boolean isAnimatedItem(int position);
    }
}

File: app/src/main/java/com/mercandalli/android/apps/files/file_list/ScaleAnimationAdapter.kt
package com.mercandalli.android.apps.files.file_list

import android.animation.Animator
import android.animation.ObjectAnimator
import android.annotation.TargetApi
import android.os.Build
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.Interpolator

/**
 * [RecyclerView] adapter with animation.
 */
@TargetApi(Build.VERSION_CODES.HONEYCOMB)
class ScaleAnimationAdapter @JvmOverloads constructor(
        /**
         * The [RecyclerView] with animated [ViewHolder].
         */
        private val recyclerView: RecyclerView,
        /**
         * The [RecyclerView.Adapter] with animated [ViewHolder].
         */
        private val mAdapter: RecyclerView.Adapter<ViewHolder>,
        /**
         * The number of [ViewHolder] items per line.
         */
        private val mViewHolderPerLine: Int = 1,
        private val mFrom: Float = 0.0f) : RecyclerView.Adapter<ViewHolder>(), Animator.AnimatorListener {

    /**
     * The animation duration for each [RecyclerView].
     */
    private var mDuration = 220

    /**
     * The animation offset duration between each [RecyclerView].
     */
    private var mOffsetDuration = 50

    /**
     * The animation [Interpolator].
     */
    private var mInterpolator: Interpolator = AccelerateDecelerateInterpolator()

    private var mLastPosition = -1
    private val mIsFirstOnly = false
    @get:Synchronized
    var counter: Int = 0
        private set
    private var mAnimsInitialized: Boolean = false

    private var mNoAnimatedPosition: NoAnimatedPosition? = null

    init {
        setHasStableIds(true)
        mAdapter.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
            override fun onChanged() {
                mLastPosition = -1
                notifyDataSetChanged()
            }

            override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
                notifyItemRangeChanged(positionStart, itemCount)
            }

            override fun onItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any?) {
                notifyItemRangeChanged(positionStart, itemCount, payload)
            }

            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                notifyItemRangeInserted(positionStart, itemCount)
            }

            override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
                notifyItemRangeRemoved(positionStart, itemCount)
            }

            override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
                notifyDataSetChanged()
            }
        })
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return this.mAdapter.onCreateViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        this.mAdapter.onBindViewHolder(holder, position)
        if (this.mIsFirstOnly && position <= this.mLastPosition) {
            clear(holder.itemView)
        } else if (mNoAnimatedPosition == null || mNoAnimatedPosition!!.isAnimatedItem(position)) {
            val animators = this.getAnimators(holder.itemView)
            if (!mAnimsInitialized) {
                for (anim in animators) {
                    anim.duration = this.mDuration.toLong()
                    anim.interpolator = this.mInterpolator
                    increaseCounter()

                    if (mViewHolderPerLine > 1 && position >= mViewHolderPerLine) {
                        anim.startDelay = (mOffsetDuration * (position - mViewHolderPerLine / 2)).toLong()
                    } else {
                        anim.startDelay = (mOffsetDuration * position).toLong()
                    }
                    anim.addListener(this)
                    anim.start()
                }
            } else {
                holder.itemView.scaleX = 1f
                holder.itemView.scaleY = 1f
            }

            this.mLastPosition = position
        } else {
            holder.itemView.scaleX = 1f
            holder.itemView.scaleY = 1f
            this.mLastPosition = position
        }
    }

    override fun getItemId(position: Int): Long {
        return mAdapter.getItemId(position)
    }

    override fun onAnimationStart(animation: Animator) {

    }

    override fun onAnimationEnd(animation: Animator) {
        if (decreaseCounter()) {
            mAnimsInitialized = true
        }
    }

    override fun onAnimationCancel(animation: Animator) {

    }

    override fun onAnimationRepeat(animation: Animator) {

    }

    @Synchronized
    fun increaseCounter() {
        counter++
    }

    override fun getItemCount(): Int {
        return this.mAdapter.itemCount
    }

    fun setDuration(duration: Int) {
        this.mDuration = duration
    }

    fun setOffsetDuration(mOffsetDuration: Int) {
        this.mOffsetDuration = mOffsetDuration
    }

    fun setInterpolator(interpolator: Interpolator) {
        this.mInterpolator = interpolator
    }

    fun setStartPosition(start: Int) {
        this.mLastPosition = start
    }

    override fun getItemViewType(position: Int): Int {
        return this.mAdapter.getItemViewType(position)
    }

    @Synchronized
    private fun decreaseCounter(): Boolean {
        counter--
        return counter == 0
    }

    private fun getAnimators(view: View): Array<Animator> {
        view.scaleX = 0f
        view.scaleY = 0f
        return arrayOf(ObjectAnimator.ofFloat(view, "scaleX", this.mFrom, 1.0f), ObjectAnimator.ofFloat(view, "scaleY", this.mFrom, 1.0f))
    }

    private fun clear(v: View) {
        ViewCompat.setAlpha(v, 1.0f)
        ViewCompat.setScaleY(v, 1.0f)
        ViewCompat.setScaleX(v, 1.0f)
        v.pivotY = (v.measuredHeight / 2).toFloat()
        ViewCompat.setPivotX(v, (v.measuredWidth / 2).toFloat())
        ViewCompat.animate(v).interpolator = null
    }

    fun setNoAnimatedPosition(noAnimatedPosition: NoAnimatedPosition?) {
        mNoAnimatedPosition = noAnimatedPosition
    }

    fun reset() {
        mLastPosition = -1
        mAnimsInitialized = false
    }

    interface NoAnimatedPosition {

        /**
         * The the item not animated.
         *
         * @param position The position to check.
         * @return True if the item is animated.
         */
        fun isAnimatedItem(position: Int): Boolean
    }
}
